### PR TITLE
WindowSwitcher: Don't use Gtk

### PR DESCRIPTION
**Pull request by leolost2605**
Slowly going towards #1839

While we're here let clutter handle layouting and don't do it ourselves.

I'm not sure whether we need a border or something here. StyleContext.render_frame was called but in the CSS I didn't find any hints to a border or something similar? Also visually it seems to be the same? Although I'm not sure my eyes are good enough to make out a difference :)

### CODE REVIEW

**Review by lenemter — CHANGES_REQUESTED** [review]
This branch is broken on my machine. It displays transparent rectangle and white rectangle at the left upper corner of the screen and randomly crashes.

**Review by tintou — DISMISSED** [review]

**Comment by leolost2605** [comment]
Sorry I shuffled some things around and didn't test again 😅
Should be fixed now

**Review by lenemter — CHANGES_REQUESTED** [review]
It still crashes with the following backtrace when I try to open multitasking view:

```
(gdb) bt
#0  _cogl_offscreen_new_with_texture_full.constprop.0 (texture=0x0, flags=0, level=0) at ../cogl/cogl/cogl-offscreen.c:51
#1  0x000056303410eff7 in gala_blur_effect_update_actor_fbo (self=0x563034b3b6c0, width=-2147483648, height=-2147483648, downscale_factor=1) at ../src/Background/BlurEffect.vala:158
#2  0x000056303410f24e in gala_blur_effect_update_framebuffers (self=0x563034b3b6c0, paint_context=0x56303597db30, actor_box=0x7fffce15e200) at ../src/Background/BlurEffect.vala:171
#3  0x00005630341100a2 in gala_blur_effect_real_paint_node (base=0x563034b3b6c0, node=0x56303609c680, paint_context=0x56303597db30, flags=CLUTTER_EFFECT_PAINT_ACTOR_DIRTY)
    at ../src/Background/BlurEffect.vala:279
#4  0x00007f174d4fd8e7 in clutter_effect_real_paint (effect=0x563034b3b6c0, node=0x56303609c680, paint_context=0x56303597db30, flags=CLUTTER_EFFECT_PAINT_ACTOR_DIRTY)
    at ../clutter/clutter/clutter-effect.c:243
#5  0x00007f174d4d4b37 in _clutter_effect_paint (flags=CLUTTER_EFFECT_PAINT_ACTOR_DIRTY, paint_context=0x56303597db30, node=0x56303609c680, effect=0x563034b3b6c0)
    at ../clutter/clutter/clutter-effect.c:307
#6  clutter_actor_continue_paint (self=<optimized out>, paint_context=0x56303597db30) at ../clutter/clutter/clutter-actor.c:3962
#7  0x00007f174d519bea in clutter_paint_node_paint (node=0x5630357fa300, paint_context=0x56303597db30) at ../clutter/clutter/clutter-paint-node.c:1080
#8  0x00007f174d519c03 in clutter_paint_node_paint (node=0x56303572cec0, paint_context=0x56303597db30) at ../clutter/clutter/clutter-paint-node.c:1087
#9  0x00007f174d4d4133 in clutter_actor_paint (self=self@entry=0x563034be4720, paint_context=paint_context@entry=0x56303597db30) at ../clutter/clutter/clutter-actor.c:3866
#10 0x00007f174d4d49ab in clutter_actor_real_paint (actor=<optimized out>, paint_context=0x56303597db30) at ../clutter/clutter/clutter-actor.c:3567
#11 0x00007f174d519bea in clutter_paint_node_paint (node=0x5630357f7c00, paint_context=0x56303597db30) at ../clutter/clutter/clutter-paint-node.c:1080
#12 0x00007f174d519c03 in clutter_paint_node_paint (node=0x56303609c630, paint_context=0x56303597db30) at ../clutter/clutter/clutter-paint-node.c:1087
#13 0x00007f174d519c03 in clutter_paint_node_paint (node=0x56303572cd10, paint_context=0x56303597db30) at ../clutter/clutter/clutter-paint-node.c:1087
#14 0x00007f174d4d4133 in clutter_actor_paint (self=self@entry=0x563034b808c0, paint_context=paint_context@entry=0x56303597db30) at ../clutter/clutter/clutter-actor.c:3866
#15 0x00007f174d4d49ab in clutter_actor_real_paint (actor=<optimized out>, paint_context=0x56303597db30) at ../clutter/clutter/clutter-actor.c:3567
#16 0x00007f174d519bea in clutter_paint_node_paint (node=0x563034eb38c0, paint_context=0x56303597db30) at ../clutter/clutter/clutter-paint-node.c:1080
#17 0x00007f174d4d4133 in clutter_actor_paint (self=self@entry=0x563034b62cc0, paint_context=paint_context@entry=0x56303597db30) at ../clutter/clutter/clutter-actor.c:3866
#18 0x00007f174d4d49ab in clutter_actor_real_paint (actor=<optimized out>, paint_context=0x56303597db30) at ../clutter/clutter/clutter-actor.c:3567
#19 0x00007f174d52a078 in clutter_stage_paint (actor=0x5630344ffda0, paint_context=0x56303597db30) at ../clutter/clutter/clutter-stage.c:1256
#20 0x00007f174d289742 in meta_stage_paint (actor=0x5630344ffda0, paint_context=0x56303597db30) at ../src/backends/meta-stage.c:212
#21 0x00007f174d519bea in clutter_paint_node_paint (node=0x5630357f7cc0, paint_context=0x56303597db30) at ../clutter/clutter/clutter-paint-node.c:1080
#22 0x00007f174d519c03 in clutter_paint_node_paint (node=0x56303572cda0, paint_context=0x56303597db30) at ../clutter/clutter/clutter-paint-node.c:1087
#23 0x00007f174d4d4133 in clutter_actor_paint (self=<optimized out>, paint_context=<optimized out>) at ../clutter/clutter/clutter-actor.c:3866
#24 0x00007f174d52aaff in clutter_stage_do_paint_view (stage=0x5630344ffda0, view=0x5630347bf0f0, redraw_clip=0x56303598a890) at ../clutter/clutter/clutter-stage.c:436
#25 0x00007f174d28458e in meta_stage_paint_view (stage=0x5630344ffda0, view=0x5630347bf0f0, redraw_clip=0x56303598a890) at ../src/backends/meta-stage.c:259
#26 0x00007f174d525134 in clutter_stage_paint_view (redraw_clip=0x56303598a890, view=0x5630347bf0f0, stage=0x5630344ffda0) at ../clutter/clutter/clutter-stage.c:459
#27 clutter_stage_paint_view (stage=stage@entry=0x5630344ffda0, view=view@entry=0x5630347bf0f0, redraw_clip=redraw_clip@entry=0x56303598a890) at ../clutter/clutter/clutter-stage.c:444
#28 0x00007f174d289d5a in paint_stage (stage_impl=<optimized out>, stage_impl=0x5630347958f0, redraw_clip=0x56303598a890, stage_view=0x5630347bf0f0)
    at ../src/backends/meta-stage-impl.c:414
#29 meta_stage_impl_redraw_view_primary (frame=0x7fffce15eeb0, stage_view=<optimized out>, stage_impl=0x5630347958f0) at ../src/backends/meta-stage-impl.c:623
#30 meta_stage_impl_redraw_view (stage_window=<optimized out>, stage_view=<optimized out>, frame=0x7fffce15eeb0) at ../src/backends/meta-stage-impl.c:748
#31 0x00007f174d298332 in meta_stage_x11_redraw_view (stage_window=<optimized out>, view=<optimized out>, frame=0x7fffce15eeb0) at ../src/backends/x11/meta-stage-x11.c:512
#32 0x00007f174d5312c8 in _clutter_stage_window_redraw_view (frame=0x7fffce15eeb0, view=0x5630347bf0f0, window=0x5630347958f0) at ../clutter/clutter/clutter-stage-window.c:94
#33 handle_frame_clock_frame (frame_clock=0x5630347c0000, frame_count=<optimized out>, user_data=0x5630347bf0f0) at ../clutter/clutter/clutter-stage-view.c:1192
#34 0x00007f174d5040e9 in clutter_frame_clock_dispatch (time_us=322686866, frame_clock=0x5630347c0000) at ../clutter/clutter/clutter-frame-clock.c:789
#35 frame_clock_source_dispatch (source=<optimized out>, callback=<optimized out>, user_data=<optimized out>) at ../clutter/clutter/clutter-frame-clock.c:845
#36 0x00007f174e39ad3b in g_main_dispatch (context=0x5630343f9240) at ../../../glib/gmain.c:3419
#37 g_main_context_dispatch (context=0x5630343f9240) at ../../../glib/gmain.c:4137
#38 0x00007f174e3f0258 in g_main_context_iterate.constprop.0 (context=0x5630343f9240, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../../../glib/gmain.c:4213
#39 0x00007f174e39a2b3 in g_main_loop_run (loop=0x5630348dc5b0) at ../../../glib/gmain.c:4413
#40 0x00007f174d2d0b79 in meta_context_run_main_loop (context=<optimized out>, error=0x7fffce15f150) at ../src/core/meta-context.c:437
#41 0x00005630340c9129 in gala_main (args=0x7fffce15f788, args_length1=1) at ../src/Main.vala:85
#42 0x00005630340c9314 in main (argc=2, argv=0x7fffce15f788) at ../src/Main.vala:29
```

And it window switcher doesn't display anything. Probably related:

```
(gala:4749): Clutter-WARNING **: 21:46:21.858: Can't update stage views actor <unnamed>[<MetaStage>:0x5615fd6c8bf0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:21.859: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fdd2ecc0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:21.859: Can't update stage views actor <unnamed>[<GalaWindowSwitcher>:0x5615fdd3ac80] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:21.859: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fe8f2ca0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:21.859: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615feb62350] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:21.859: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615feb509c0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:21.859: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615feb626b0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:21.859: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615feb50cf0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:22.976: Can't update stage views actor <unnamed>[<MetaStage>:0x5615fd6c8bf0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:22.976: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fdd2ecc0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:22.976: Can't update stage views actor <unnamed>[<GalaWindowSwitcher>:0x5615fdd3ac80] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:22.976: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fe8f2ca0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:22.976: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615feb62a10] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:22.976: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615feb51020] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:22.976: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615feb62d70] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:22.976: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615feb51350] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.012: Can't update stage views actor <unnamed>[<MetaStage>:0x5615fd6c8bf0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.012: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fdd2ecc0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.012: Can't update stage views actor <unnamed>[<GalaWindowSwitcher>:0x5615fdd3ac80] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.012: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fe8f2ca0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.012: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615feb630d0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.012: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615feb51680] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.012: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615feb63430] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.012: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615feb519b0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.677: Can't update stage views actor <unnamed>[<MetaStage>:0x5615fd6c8bf0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.677: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fdd2ecc0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.678: Can't update stage views actor <unnamed>[<GalaWindowSwitcher>:0x5615fdd3ac80] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.678: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fe8f2ca0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.678: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615feb63790] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.678: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615feb51ce0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.678: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615feb63af0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:24.678: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615ff2b2490] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:28.545: Can't update stage views actor <unnamed>[<MetaStage>:0x5615fd6c8bf0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:28.545: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fdd2ecc0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:28.545: Can't update stage views actor <unnamed>[<GalaWindowSwitcher>:0x5615fdd3ac80] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:28.545: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fe8f2ca0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:28.545: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615feb63e50] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:28.545: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615ff2b27c0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:28.545: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615ff282420] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:28.545: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615ff2b2af0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:31.845: Can't update stage views actor <unnamed>[<MetaStage>:0x5615fd6c8bf0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:31.845: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fdd2ecc0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:31.845: Can't update stage views actor <unnamed>[<GalaWindowSwitcher>:0x5615fdd3ac80] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:31.845: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fe8f2ca0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:31.845: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615ff282780] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:31.845: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615ff2b2e20] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:31.845: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615ff282ae0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:31.845: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615ff2b3150] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:35.845: Can't update stage views actor <unnamed>[<MetaStage>:0x5615fd6c8bf0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:35.845: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fdd2ecc0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:35.845: Can't update stage views actor <unnamed>[<GalaWindowSwitcher>:0x5615fdd3ac80] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:35.845: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fe8f2ca0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:35.845: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615ff282e40] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:35.845: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615ff2b3480] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:35.845: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615ff2831a0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:35.845: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615ff2b37b0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:43.188: Can't update stage views actor <unnamed>[<MetaStage>:0x5615fd6c8bf0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:43.188: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fdd2ecc0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:43.188: Can't update stage views actor <unnamed>[<GalaWindowSwitcher>:0x5615fdd3ac80] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:43.188: Can't update stage views actor <unnamed>[<ClutterActor>:0x5615fe8f2ca0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:43.188: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615ff283500] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:43.188: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615ff2b3ae0] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:43.188: Can't update stage views actor <unnamed>[<GalaWindowSwitcherIcon>:0x5615ff283860] is on because it needs an allocation.

(gala:4749): Clutter-WARNING **: 21:46:43.188: Can't update stage views actor <unnamed>[<GalaWindowIcon>:0x5615ff2b3e10] is on because it needs an allocation.
```

Nothing like this happens in `master` but the backtrace just seems to be completely unrelated to the window switcher ¯\\\_(ツ)\_/¯

**Comment by leolost2605** [comment]
Ok I fixed another crash I hope that was it 🤞

File: src/Widgets/WindowSwitcher/WindowSwitcher.vala [hunk 1]
```diff
@@ -64,114 +61,112 @@ public class Gala.WindowSwitcher : CanvasActor {
...
+            layout_manager = new Clutter.FlowLayout (Clutter.FlowOrientation.HORIZONTAL)
+        };
 
-        opacity = 0;
+        caption = new Clutter.Text () {
+            font_name = CAPTION_FONT_NAME,
+            ellipsize = END,
+            line_alignment = CENTER
+        };
+        caption.set_pivot_point (0.5f, 0.5f);
```
Comment (lenemter) [inline src/Widgets/WindowSwitcher/WindowSwitcher.vala:74]: While we're here can you remove pivot point here. We don't scale the caption, so no reason to set it manually.

**Review by lenemter — APPROVED** [review]
Doesn't crash anymore and works as expected. Good job!

**Review by danirabbit — APPROVED** [review]
Just confirming everything looks the same here to my eyes. I think `render_frame` is supposed to render all box shadows IIRC. There's a missing inner box shadow but this is also the case in master so LGTM :rocket:

File: src/Widgets/WindowSwitcher/WindowSwitcher.vala (hunk 1 above)
Comment (leolost2605) [reply src/Widgets/WindowSwitcher/WindowSwitcher.vala]: Will do, I just wasn't quite sure what it was used for so I left it there :)

**Pull request merged**